File: lib/flash_profile/cost.ex
defmodule FlashProfile.Cost do
  @moduledoc """
  Cost function for FlashProfile patterns.

  The cost balances pattern specificity vs simplicity using:
  - Static cost: intrinsic cost of each atom (from atom definition)
  - Dynamic weight: fraction of string length matched by each atom

  Lower cost = better pattern.

  ## Cost Formula

  From the FlashProfile paper (Section 4.3):

  ```
  C_FP(P, S) = Σ Q(αi) · W(i, S | P)
  ```

  Where:
  - P = [α1, α2, ..., αk] is a pattern (list of atoms)
  - Q(αi) is the static cost of atom αi
  - W(i, S | P) is the dynamic weight for atom i

  ## Dynamic Weight Formula

  ```
  W(i, S | P) = (1/|S|) · Σ_{s∈S} (αi(si) / |s|)
  ```

  Where:
  - s1 = s (the original string)
  - si+1 = si[αi(si):] (remaining suffix after matching atom αi)
  - αi(si) is the length matched by atom i on string si
  - |s| is the total length of the original string

  The dynamic weight is the average fraction of the string length matched
  by that atom across all strings in S.

  ## Cost Meaning

  - Higher cost = less desirable pattern
  - Lower cost = more desirable pattern
  - Empty pattern on empty string set has cost 0
  - If pattern doesn't match all strings, cost is :infinity
  """

  alias FlashProfile.Atom

  @type cost :: float() | :infinity
  @type pattern :: [Atom.t()]

  @doc """
  Calculate the cost of a pattern over a dataset.

  Returns :infinity if pattern doesn't match all strings in the dataset.
  Returns 0.0 for empty pattern on empty dataset.

  ## Parameters

  - `pattern` - List of atoms forming the pattern
  - `strings` - List of strings to evaluate the pattern against

  ## Returns

  - `float()` - The cost of the pattern
  - `:infinity` - If the pattern doesn't match all strings

  ## Examples

      iex> alias FlashProfile.{Cost, Atom}
      iex> upper = Atom.char_class("Upper", ?A..?Z |> Enum.to_list(), 8.2)
      iex> lower = Atom.char_class("Lower", ?a..?z |> Enum.to_list(), 9.1)
      iex> pattern = [upper, lower]
      iex> strings = ["Male", "Female"]
      iex> cost = Cost.calculate(pattern, strings)
      iex> is_float(cost) and cost > 0
      true

      iex> alias FlashProfile.{Cost, Atom}
      iex> Cost.calculate([], [])
      0.0

      iex> alias FlashProfile.{Cost, Atom}
      iex> digit = Atom.char_class("Digit", ?0..?9 |> Enum.to_list(), 8.2)
      iex> Cost.calculate([digit], ["abc"])
      :infinity
  """
  @spec calculate(pattern(), [String.t()]) :: cost()
  def calculate([], []), do: 0.0
  def calculate([], _strings), do: :infinity
  def calculate(_pattern, []), do: 0.0

  def calculate(pattern, strings) do
    case get_all_match_lengths(pattern, strings) do
      {:error, _} ->
        :infinity

      {:ok, all_lengths} ->
        # all_lengths is list of lists: [[len1, len2, ...], [len1, len2, ...], ...]
        # Each inner list is the lengths for one string

        # Calculate cost for each atom position
        pattern
        |> Enum.with_index()
        |> Enum.map(fn {atom, idx} ->
          static_cost = Atom.static_cost(atom)
          dynamic_weight = calculate_dynamic_weight(all_lengths, strings, idx)
          static_cost * dynamic_weight
        end)
        |> Enum.sum()
    end
  end

  @doc """
  Calculate cost with detailed breakdown per atom.

  Returns {:ok, {total_cost, breakdown}} or {:error, reason} where breakdown
  is a list of tuples containing {atom, static_cost, dynamic_weight} for each
  atom in the pattern.

  ## Parameters

  - `pattern` - List of atoms forming the pattern
  - `strings` - List of strings to evaluate the pattern against

  ## Returns

  - `{:ok, {total_cost, breakdown}}` - Success with detailed breakdown
  - `{:error, reason}` - Pattern doesn't match all strings

  ## Examples

      iex> alias FlashProfile.{Cost, Atom}
      iex> digit = Atom.char_class("Digit", ?0..?9 |> Enum.to_list(), 8.2)
      iex> {:ok, {cost, breakdown}} = Cost.calculate_detailed([digit], ["123"])
      iex> is_float(cost) and length(breakdown) == 1
      true
  """
  @spec calculate_detailed(pattern(), [String.t()]) ::
          {:ok, {float(), [{Atom.t(), float(), float()}]}} | {:error, term()}
  def calculate_detailed([], []), do: {:ok, {0.0, []}}
  def calculate_detailed([], _strings), do: {:error, :empty_pattern_non_empty_strings}
  def calculate_detailed(_pattern, []), do: {:ok, {0.0, []}}

  def calculate_detailed(pattern, strings) do
    case get_all_match_lengths(pattern, strings) do
      {:error, reason} ->
        {:error, reason}

      {:ok, all_lengths} ->
        breakdown =
          pattern
          |> Enum.with_index()
          |> Enum.map(fn {atom, idx} ->
            static_cost = Atom.static_cost(atom)
            dynamic_weight = calculate_dynamic_weight(all_lengths, strings, idx)
            {atom, static_cost, dynamic_weight}
          end)

        total_cost =
          breakdown
          |> Enum.map(fn {_atom, static_cost, dynamic_weight} ->
            static_cost * dynamic_weight
          end)
          |> Enum.sum()

        {:ok, {total_cost, breakdown}}
    end
  end

  @doc """
  Compare two patterns by cost over the same dataset.

  Returns :lt if pattern1 has lower cost (is better), :eq if equal,
  or :gt if pattern2 has lower cost.

  ## Parameters

  - `pattern1` - First pattern to compare
  - `pattern2` - Second pattern to compare
  - `strings` - Dataset to evaluate both patterns against

  ## Returns

  - `:lt` - pattern1 is better (lower cost)
  - `:eq` - patterns have equal cost
  - `:gt` - pattern2 is better (lower cost)

  ## Examples

      iex> alias FlashProfile.{Cost, Atom}
      iex> digit = Atom.char_class("Digit", ?0..?9 |> Enum.to_list(), 8.2)
      iex> alpha = Atom.char_class("Alpha", (?a..?z |> Enum.to_list()) ++ (?A..?Z |> Enum.to_list()), 15.0)
      iex> Cost.compare([digit], [alpha], ["123"])
      :lt
  """
  @spec compare(pattern(), pattern(), [String.t()]) :: :lt | :eq | :gt
  def compare(pattern1, pattern2, strings) do
    cost1 = calculate(pattern1, strings)
    cost2 = calculate(pattern2, strings)

    cond do
      cost1 == :infinity and cost2 == :infinity -> :eq
      cost1 == :infinity -> :gt
      cost2 == :infinity -> :lt
      cost1 < cost2 -> :lt
      cost1 > cost2 -> :gt
      true -> :eq
    end
  end

  @doc """
  Find the minimum cost pattern from a list.

  Returns {pattern, cost} or nil if list is empty.

  ## Parameters

  - `patterns` - List of patterns to evaluate
  - `strings` - Dataset to evaluate patterns against

  ## Returns

  - `{pattern, cost}` - Best pattern and its cost
  - `nil` - If patterns list is empty

  ## Examples

      iex> alias FlashProfile.{Cost, Atom}
      iex> digit = Atom.char_class("Digit", ?0..?9 |> Enum.to_list(), 8.2)
      iex> alpha = Atom.char_class("Alpha", (?a..?z |> Enum.to_list()) ++ (?A..?Z |> Enum.to_list()), 15.0)
      iex> {best, _cost} = Cost.min_cost([[digit], [alpha]], ["123"])
      iex> best == [digit]
      true
  """
  @spec min_cost([pattern()], [String.t()]) :: {pattern(), cost()} | nil
  def min_cost([], _strings), do: nil

  def min_cost(patterns, strings) do
    patterns
    |> Enum.map(fn pattern ->
      {pattern, calculate(pattern, strings)}
    end)
    |> Enum.min_by(
      fn {_pattern, cost} ->
        case cost do
          :infinity -> :infinity
          c when is_float(c) -> c
        end
      end,
      fn
        :infinity, :infinity -> true
        :infinity, _ -> false
        _, :infinity -> true
        c1, c2 -> c1 <= c2
      end
    )
  end

  ## Private Helper Functions

  # Get match lengths for a pattern across all strings.
  # Returns {:ok, list of length lists} or {:error, :no_match}
  #
  # For each string, returns a list of lengths matched by each atom.
  # Example: pattern [Upper, Lower] on "Male" returns [1, 3]
  # (Upper matches "M" = 1 char, Lower matches "ale" = 3 chars)
  @spec get_all_match_lengths(pattern(), [String.t()]) ::
          {:ok, [[non_neg_integer()]]} | {:error, :no_match}
  defp get_all_match_lengths(pattern, strings) do
    results =
      Enum.map(strings, fn string ->
        match_pattern_lengths(pattern, string)
      end)

    if Enum.any?(results, &is_nil/1) do
      {:error, :no_match}
    else
      {:ok, results}
    end
  end

  # Match a pattern against a string and return the list of lengths
  # matched by each atom, or nil if the pattern doesn't fully match.
  #
  # The pattern must consume the entire string for a successful match.
  @spec match_pattern_lengths(pattern(), String.t()) :: [non_neg_integer()] | nil
  defp match_pattern_lengths(pattern, string) do
    {lengths, remaining} =
      Enum.reduce(pattern, {[], string}, fn atom, {lengths_acc, str} ->
        length = Atom.match(atom, str)

        if length > 0 do
          # Atom matched, consume that portion of the string
          new_remaining = String.slice(str, length..-1//1)
          {lengths_acc ++ [length], new_remaining}
        else
          # Atom didn't match, stop processing
          {lengths_acc ++ [0], str}
        end
      end)

    # Pattern must match entire string (no remainder)
    # and all atoms must have matched (no zeros in lengths)
    if remaining == "" and Enum.all?(lengths, &(&1 > 0)) do
      lengths
    else
      nil
    end
  end

  # Calculate the dynamic weight for a specific atom position.
  #
  # W(i, S | P) = (1/|S|) · Σ_{s∈S} (αi(si) / |s|)
  #
  # This is the average fraction of the original string length that
  # this atom matches across all strings.
  @spec calculate_dynamic_weight([[non_neg_integer()]], [String.t()], non_neg_integer()) ::
          float()
  defp calculate_dynamic_weight(all_lengths, strings, atom_index) do
    num_strings = length(strings)

    if num_strings == 0 do
      0.0
    else
      total =
        Enum.zip(all_lengths, strings)
        |> Enum.map(fn {lengths, string} ->
          atom_length = Enum.at(lengths, atom_index, 0)
          string_length = String.length(string)

          if string_length > 0 do
            atom_length / string_length
          else
            0.0
          end
        end)
        |> Enum.sum()

      total / num_strings
    end
  end
end

File: native/flash_profile/cost.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const atom_mod = @import("atom.zig");
const Atom = atom_mod.Atom;
const pattern_mod = @import("pattern.zig");
const Pattern = pattern_mod.Pattern;
const types = @import("types.zig");
const Cost = types.Cost;

/// Calculate the FlashProfile cost function for a pattern over a dataset
///
/// From the paper (Section 4.3):
/// C_FP(P, S) = Σ Q(αi) · W(i, S | P)
///
/// Where:
/// - Q(αi) is the static cost of atom i
/// - W(i, S | P) is the dynamic weight (average fraction of string consumed by atom i)
/// - len_i(s) is the length matched by atom i on string s
///
/// Returns:
/// - Cost.infinity if pattern doesn't match all strings
/// - Cost.finite(0.0) for empty pattern on empty dataset
/// - Cost.finite(0.0) for any pattern on empty dataset (vacuous truth)
/// - Cost.infinity for empty pattern on non-empty dataset
///
/// ## Edge Cases
///
/// - **Empty pattern on empty dataset**: Returns 0.0 (vacuous truth)
/// - **Empty strings in dataset**: Empty strings contribute 0.0 to the dynamic weight
///   calculation (avoiding division by zero). This means empty strings effectively
///   don't affect the cost.
/// - **Pattern on empty dataset**: Returns 0.0 (any pattern vacuously matches empty set)
pub fn calculateCost(
    pattern: []const Atom,
    strings: []const []const u8,
    allocator: Allocator,
) !Cost {
    // Edge case: empty pattern on empty dataset
    if (pattern.len == 0 and strings.len == 0) {
        return Cost.fromFinite(0.0);
    }

    // Edge case: empty pattern on non-empty dataset = infinity
    if (pattern.len == 0 and strings.len > 0) {
        return Cost.asInfinity();
    }

    // Edge case: pattern on empty dataset = 0.0 (vacuous truth)
    if (strings.len == 0) {
        return Cost.fromFinite(0.0);
    }

    // Get match lengths for all strings
    var all_lengths = try getAllMatchLengths(pattern, strings, allocator);
    defer {
        for (all_lengths.items) |lengths| {
            allocator.free(lengths);
        }
        all_lengths.deinit(allocator);
    }

    // Check if all strings matched (all_lengths entries are non-null)
    var all_matched = true;
    for (all_lengths.items) |lengths| {
        if (lengths.len == 0) {
            all_matched = false;
            break;
        }
    }

    if (!all_matched) {
        return Cost.asInfinity();
    }

    // Calculate cost as sum over atoms
    var total_cost: f64 = 0.0;

    for (pattern, 0..) |atom_val, idx| {
        const static_cost = atom_val.static_cost;
        const dynamic_weight = calculateDynamicWeight(all_lengths.items, strings, idx);
        total_cost += static_cost * dynamic_weight;
    }

    return Cost.fromFinite(total_cost);
}

/// Calculate the dynamic weight for atom at position `atom_index`
///
/// W(i, S | P) = (1/|S|) · Σ_{s∈S} (αi(si) / |s|)
///
/// This is the average fraction of the original string length that
/// this atom matches across all strings.
///
/// Where:
/// - s1 = s (the original string)
/// - si+1 = si[αi(si):] (remaining suffix after matching atom αi)
/// - αi(si) is the length matched by atom i on string si
/// - |s| is the total length of the original string
pub fn dynamicWeight(
    position: usize,
    pattern: []const Atom,
    strings: []const []const u8,
    allocator: Allocator,
) !f64 {
    if (strings.len == 0) {
        return 0.0;
    }

    // Get match lengths for all strings
    var all_lengths = try getAllMatchLengths(pattern, strings, allocator);
    defer {
        for (all_lengths.items) |lengths| {
            allocator.free(lengths);
        }
        all_lengths.deinit(allocator);
    }

    return calculateDynamicWeight(all_lengths.items, strings, position);
}

/// Calculate dynamic weight from pre-computed match lengths
fn calculateDynamicWeight(
    all_lengths: []const []const usize,
    strings: []const []const u8,
    atom_index: usize,
) f64 {
    const num_strings = strings.len;
    if (num_strings == 0) {
        return 0.0;
    }

    var total: f64 = 0.0;

    for (all_lengths, 0..) |lengths, i| {
        // Get the length matched by this atom
        const atom_length = if (atom_index < lengths.len)
            lengths[atom_index]
        else
            0;

        const string_length = strings[i].len;

        if (string_length > 0) {
            const fraction = @as(f64, @floatFromInt(atom_length)) / @as(f64, @floatFromInt(string_length));
            total += fraction;
        }
        // Empty strings contribute 0.0
    }

    return total / @as(f64, @floatFromInt(num_strings));
}

/// Get match lengths for a pattern across all strings
///
/// Returns ArrayList of length arrays. Each inner array contains the lengths
/// matched by each atom for one string. If a string doesn't match, returns
/// an empty array for that string.
fn getAllMatchLengths(
    pattern: []const Atom,
    strings: []const []const u8,
    allocator: Allocator,
) !std.ArrayList([]const usize) {
    var results: std.ArrayList([]const usize) = .{};
    errdefer {
        for (results.items) |lengths| {
            allocator.free(lengths);
        }
        results.deinit(allocator);
    }

    const pat = Pattern.init(pattern);

    for (strings) |string| {
        const lengths = try pat.matchLengths(allocator, string);
        if (lengths) |l| {
            try results.append(allocator, l);
        } else {
            // Pattern doesn't match - append empty array
            const empty = try allocator.alloc(usize, 0);
            try results.append(allocator, empty);
        }
    }

    return results;
}

/// Cost breakdown for a single atom
pub const CostBreakdown = struct {
    atom_index: usize,
    static_cost: f64,
    dynamic_weight: f64,
    contribution: f64,
};

/// Result of detailed cost calculation
pub const DetailedCostResult = struct {
    total_cost: f64,
    breakdown: []CostBreakdown,
    allocator: Allocator,

    pub fn deinit(self: *DetailedCostResult) void {
        self.allocator.free(self.breakdown);
    }
};

/// Calculate cost with detailed breakdown for each atom
///
/// Returns null if pattern doesn't match all strings in the dataset.
///
/// ## Parameters
///
/// - `pattern` - List of atoms forming the pattern
/// - `strings` - List of strings to evaluate the pattern against
/// - `allocator` - Memory allocator
///
/// ## Returns
///
/// - `DetailedCostResult` - Success with detailed breakdown
/// - `null` - Pattern doesn't match all strings
pub fn calculateCostDetailed(
    pattern: []const Atom,
    strings: []const []const u8,
    allocator: Allocator,
) !?DetailedCostResult {
    // Edge cases
    if (pattern.len == 0 and strings.len == 0) {
        const breakdown = try allocator.alloc(CostBreakdown, 0);
        return DetailedCostResult{
            .total_cost = 0.0,
            .breakdown = breakdown,
            .allocator = allocator,
        };
    }

    if (pattern.len == 0 and strings.len > 0) {
        return null; // Error: empty pattern on non-empty strings
    }

    if (strings.len == 0) {
        const breakdown = try allocator.alloc(CostBreakdown, 0);
        return DetailedCostResult{
            .total_cost = 0.0,
            .breakdown = breakdown,
            .allocator = allocator,
        };
    }

    // Get match lengths
    var all_lengths = try getAllMatchLengths(pattern, strings, allocator);
    defer {
        for (all_lengths.items) |lengths| {
            allocator.free(lengths);
        }
        all_lengths.deinit(allocator);
    }

    // Check if all strings matched
    for (all_lengths.items) |lengths| {
        if (lengths.len == 0) {
            return null; // Pattern doesn't match all strings
        }
    }

    // Build breakdown
    const breakdown = try allocator.alloc(CostBreakdown, pattern.len);
    errdefer allocator.free(breakdown);

    var total_cost: f64 = 0.0;

    for (pattern, 0..) |atom_val, i| {
        const static_cost = atom_val.static_cost;
        const dynamic_weight = calculateDynamicWeight(all_lengths.items, strings, i);
        const contribution = static_cost * dynamic_weight;

        breakdown[i] = .{
            .atom_index = i,
            .static_cost = static_cost,
            .dynamic_weight = dynamic_weight,
            .contribution = contribution,
        };

        total_cost += contribution;
    }

    return DetailedCostResult{
        .total_cost = total_cost,
        .breakdown = breakdown,
        .allocator = allocator,
    };
}

// ============================================================================
// Tests
// ============================================================================

test "cost: empty pattern on empty dataset" {
    const testing = std.testing;
    const allocator = testing.allocator;

    const pattern = [_]Atom{};
    const strings = [_][]const u8{};

    const cost = try calculateCost(&pattern, &strings, allocator);
    try testing.expect(cost == .finite);
    try testing.expectEqual(@as(f64, 0.0), cost.finite);
}

test "cost: empty pattern on non-empty dataset" {
    const testing = std.testing;
    const allocator = testing.allocator;

    const pattern = [_]Atom{};
    const strings = [_][]const u8{"test"};

    const cost = try calculateCost(&pattern, &strings, allocator);
    try testing.expect(cost == .infinity);
}

test "cost: pattern on empty dataset" {
    const testing = std.testing;
    const allocator = testing.allocator;

    const d = atom_mod.digit();
    const pattern = [_]Atom{d};
    const strings = [_][]const u8{};

    const cost = try calculateCost(&pattern, &strings, allocator);
    try testing.expect(cost == .finite);
    try testing.expectEqual(@as(f64, 0.0), cost.finite);
}

test "cost: simple pattern matching" {
    const testing = std.testing;
    const allocator = testing.allocator;

    const u = atom_mod.upper();
    const d = atom_mod.digit();
    const pattern = [_]Atom{ u, d };
    const strings = [_][]const u8{ "A123", "B456" };

    const cost = try calculateCost(&pattern, &strings, allocator);
    try testing.expect(cost == .finite);
    try testing.expect(cost.finite > 0.0);
}

test "cost: pattern doesn't match" {
    const testing = std.testing;
    const allocator = testing.allocator;

    const d = atom_mod.digit();
    const pattern = [_]Atom{d};
    const strings = [_][]const u8{"abc"};

    const cost = try calculateCost(&pattern, &strings, allocator);
    try testing.expect(cost == .infinity);
}

test "cost: dynamic weight calculation" {
    const testing = std.testing;
    const allocator = testing.allocator;

    const u = atom_mod.upper();
    const d = atom_mod.digit();
    const pattern = [_]Atom{ u, d };

    // "A123" -> Upper matches 1 char (1/4 = 0.25), Digit matches 3 chars (3/4 = 0.75)
    const strings = [_][]const u8{"A123"};

    const weight0 = try dynamicWeight(0, &pattern, &strings, allocator);
    const weight1 = try dynamicWeight(1, &pattern, &strings, allocator);

    try testing.expectApproxEqAbs(@as(f64, 0.25), weight0, 0.001);
    try testing.expectApproxEqAbs(@as(f64, 0.75), weight1, 0.001);
}

test "cost: detailed breakdown" {
    const testing = std.testing;
    const allocator = testing.allocator;

    const d = atom_mod.digit();
    const pattern = [_]Atom{d};
    const strings = [_][]const u8{"123"};

    var result = try calculateCostDetailed(&pattern, &strings, allocator);
    defer if (result) |*r| r.deinit();

    try testing.expect(result != null);
    try testing.expectEqual(@as(usize, 1), result.?.breakdown.len);
    try testing.expectEqual(@as(f64, 1.0), result.?.breakdown[0].dynamic_weight);
}

test "cost: complex multi-atom pattern" {
    const testing = std.testing;
    const allocator = testing.allocator;

    // Test cost calculation for pattern with 4+ atoms
    // Pattern: Upper + Digit + DotDash + Lower + Digit
    const u = atom_mod.upper();
    const d = atom_mod.digit();
    const dd = atom_mod.dotDash();
    const l = atom_mod.lower();
    const pattern = [_]Atom{ u, d, dd, l, d };

    // Test strings matching the pattern: "A1-b2", "Z9.x5"
    const strings = [_][]const u8{ "A1-b2", "Z9.x5" };

    const cost = try calculateCost(&pattern, &strings, allocator);
    try testing.expect(cost == .finite);
    try testing.expect(cost.finite > 0.0);

    // Verify detailed breakdown works correctly
    var result = try calculateCostDetailed(&pattern, &strings, allocator);
    defer if (result) |*r| r.deinit();

    try testing.expect(result != null);
    try testing.expectEqual(@as(usize, 5), result.?.breakdown.len);

    // Each atom should contribute to the total cost
    var sum: f64 = 0.0;
    for (result.?.breakdown) |item| {
        try testing.expect(item.contribution >= 0.0);
        sum += item.contribution;
    }
    try testing.expectApproxEqAbs(result.?.total_cost, sum, 0.001);
}

test "cost: very long strings" {
    const testing = std.testing;
    const allocator = testing.allocator;

    // Test cost calculation with strings of 100+ characters
    // Verify no overflow or precision issues
    const d = atom_mod.digit();
    const l = atom_mod.lower();
    const pattern = [_]Atom{ d, l };

    // Create long strings (100+ chars)
    // Format: "1" + 100 'a's + "2" + 100 'b's
    var string1: std.ArrayList(u8) = .{};
    defer string1.deinit(allocator);
    try string1.append(allocator, '1');
    {
        var i: usize = 0;
        while (i < 100) : (i += 1) {
            try string1.append(allocator, 'a');
        }
    }

    var string2: std.ArrayList(u8) = .{};
    defer string2.deinit(allocator);
    try string2.append(allocator, '2');
    {
        var i: usize = 0;
        while (i < 100) : (i += 1) {
            try string2.append(allocator, 'b');
        }
    }

    const strings = [_][]const u8{ string1.items, string2.items };

    const cost = try calculateCost(&pattern, &strings, allocator);
    try testing.expect(cost == .finite);
    try testing.expect(cost.finite > 0.0);

    // Verify dynamic weights are reasonable (each string is 101 chars)
    const weight0 = try dynamicWeight(0, &pattern, &strings, allocator);
    const weight1 = try dynamicWeight(1, &pattern, &strings, allocator);

    // First atom (digit) matches 1 char out of 101 = ~0.0099
    try testing.expectApproxEqAbs(@as(f64, 1.0 / 101.0), weight0, 0.001);
    // Second atom (lower) matches 100 chars out of 101 = ~0.9901
    try testing.expectApproxEqAbs(@as(f64, 100.0 / 101.0), weight1, 0.001);
}

test "cost: pattern with repeated atom type" {
    const testing = std.testing;
    const allocator = testing.allocator;

    // Test pattern like [Digit, Digit, Digit]
    // Verify cost accumulation is correct
    // Use fixed-width atoms to match exactly 1 char each
    const d = atom_mod.withFixedWidth(atom_mod.digit(), 1);
    const pattern = [_]Atom{ d, d, d };

    const strings = [_][]const u8{ "123", "456", "789" };

    const cost = try calculateCost(&pattern, &strings, allocator);
    try testing.expect(cost == .finite);
    try testing.expect(cost.finite > 0.0);

    // Check detailed breakdown
    var result = try calculateCostDetailed(&pattern, &strings, allocator);
    defer if (result) |*r| r.deinit();

    try testing.expect(result != null);
    try testing.expectEqual(@as(usize, 3), result.?.breakdown.len);

    // Each digit atom should have the same static cost
    const static_cost = result.?.breakdown[0].static_cost;
    try testing.expectEqual(static_cost, result.?.breakdown[1].static_cost);
    try testing.expectEqual(static_cost, result.?.breakdown[2].static_cost);

    // Each digit matches exactly 1 char out of 3 = 1/3 dynamic weight
    for (result.?.breakdown) |item| {
        try testing.expectApproxEqAbs(@as(f64, 1.0 / 3.0), item.dynamic_weight, 0.001);
    }
}

test "cost: dynamic weight with uniform matches" {
    const testing = std.testing;
    const allocator = testing.allocator;

    // Test when all strings have same match lengths
    // Dynamic weight should be predictable
    const u = atom_mod.upper();
    const d = atom_mod.digit();
    const l = atom_mod.lower();
    const pattern = [_]Atom{ u, d, l };

    // All strings: 1 upper + 2 digits + 3 lowers = 6 chars total
    const strings = [_][]const u8{ "A12abc", "B34def", "C56ghi", "D78jkl" };

    const cost = try calculateCost(&pattern, &strings, allocator);
    try testing.expect(cost == .finite);
    try testing.expect(cost.finite > 0.0);

    // Check dynamic weights
    const weight0 = try dynamicWeight(0, &pattern, &strings, allocator);
    const weight1 = try dynamicWeight(1, &pattern, &strings, allocator);
    const weight2 = try dynamicWeight(2, &pattern, &strings, allocator);

    // Upper matches 1/6, Digit matches 2/6, Lower matches 3/6
    try testing.expectApproxEqAbs(@as(f64, 1.0 / 6.0), weight0, 0.001);
    try testing.expectApproxEqAbs(@as(f64, 2.0 / 6.0), weight1, 0.001);
    try testing.expectApproxEqAbs(@as(f64, 3.0 / 6.0), weight2, 0.001);

    // Sum of dynamic weights should equal 1.0
    const sum = weight0 + weight1 + weight2;
    try testing.expectApproxEqAbs(@as(f64, 1.0), sum, 0.001);
}
